crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -291,6 +291,10 @@ pub(super) struct TypeInferenceBuilder<'db> {
     // Cached lookups
     file: File,
 
+    /// Offsets of [`ast::ExprName`] nodes representing unbound names
+    /// for which we've already added diagnostics
+    unbound_names: FxHashSet<TextSize>,
```

Comment:
In which cases did you see duplicate diagnostics if you didn't do this?

In general it is supposed to be an invariant that we don't double-infer anything, so this _shouldn't_ be necessary. Unpacking assignment is one known counter-example currently.

If there are non-unpacking-assignment cases where you saw duplicate diagnostics, that would be interesting, as it might reveal a bug.

Unless there are so many "bugs" here that it turns out we just can't feasibly maintain this invariant, I would rather not add this and just have duplicate diagnostics in some tests, with TODO comments to fix those cases by not double-inferring the expression.

If for some reason we can't maintain the no-double-inference invariant, it means we are going to need to generalize this hashset approach to not just unbound-names, but all diagnostics.

Reply:
> In which cases did you see duplicate diagnostics if you didn't do this?

Take a look at the change to the assertions in the tomllib benchmark in this PR. Both of these diagnostics were being reported twice with the lint rule as it was previously:

```
    "/src/tomllib/_parser.py:104:14: Name `char` used when possibly not defined",
    "/src/tomllib/_parser.py:115:14: Name `char` used when possibly not defined",
```

Both of these are false positives (due to the fact our control-flow analysis doesn't yet understand terminal statements), but neither appears to be anything to do with unpacking, as far as I can see:

- https://github.com/python/cpython/blob/591f6754b56cb7f6c31fce8c22528bdf0a99556c/Lib/tomllib/_parser.py#L104
- https://github.com/python/cpython/blob/591f6754b56cb7f6c31fce8c22528bdf0a99556c/Lib/tomllib/_parser.py#L115

I'm okay with keeping the duplicate diagnostics for now if you are!

Reply:
Oh, but now I try running the benchmark again without the deduplication logic, I don't see it anymore.

Hmm, okay. I think we _did_ have duplicated diagnostics when it was implemented outside of `infer.rs`, but now it's implemented as part of `infer.rs`, we don't anymore. But for some reason I thought we still did, so tried to fix it. Argh.

Anyway, you're right, we don't need any of this deduplication logic. Hooray!